File: leetcode3/이진희/70. Climbing Stairs.java
/*

1. 아이디어 : 
   n을 1부터 4까지 직접 경우의 수를 구한 후, 수학적 규칙을 찾아 계산

2. 시간복잡도 : O(N)

3. 자료구조/알고리즘 : DP

 */

class Solution {
    public int climbStairs(int n) {

        // 1 2 3 5
        // 22 1111 211 -> 5
        
        if(n == 1) return 1;
        if(n == 2) return 2;
        
        int[] dp = new int[n+1];
        dp[1] = 1;
        dp[2] = 2;

        for(int i=3; i<=n; i++) {
            dp[i] = dp[i-1] + dp[i-2];
        }

        return dp[n];

    }
}

File: leetcode3/최원준/70. Climbing Stairs.py
#

'''
1. 아이디어 :
x에 도달하기 위해서는 x-1에서 1칸 점프 + x-2에서 1칸 점프이므로, f(x) = f(x-1) + f(x-2)
길이 2의 배열을 활용하여 공간 절약

2. 시간복잡도 :
    O(n)

3. 자료구조/알고리즘 :
dp

'''
class Solution:
    def climbStairs(self, n: int) -> int:
        if n == 1:
            return n

        ans = [1,2]
        for i in range(3, n+1):
            ans = [ans[1],ans[0] + ans[1]]
        return ans[1]

